src/storage/heap_file.c

```diff
@@ -25091,19 +25091,70 @@ heap_get_visible_version_with_repl_desync (THREAD_ENTRY * thread_p, HEAP_GET_CON
 	}
       assert (is_passive_transaction_server ());
       // Handle the page desynchronization error
-      if (context->home_page_watcher.pgptr)
-	{
-	  /* Unfix home page. */
-	  pgbuf_ordered_unfix (thread_p, &context->home_page_watcher);
-	}
 
-      if (context->fwd_page_watcher.pgptr != NULL)
+      // clean - saving state if required
+      //
+      heap_clean_get_context (thread_p, context);
+
+      // also unfix what is in cache watcher; saving the state for a future re-fix after the wait
+      //
+      VPID saved_vpid VPID_INITIALIZER;
+      PGBUF_LATCH_MODE saved_latch_mode = PGBUF_NO_LATCH;
+      bool restore_scan_cache_page_watcher = false;
+      if (is_heap_scan)
 	{
-	  /* Unfix forward page. */
-	  pgbuf_ordered_unfix (thread_p, &context->fwd_page_watcher);
+	  assert (context->scan_cache != nullptr);
+	  if (context->scan_cache->page_watcher.pgptr != nullptr)
+	    {
+	      assert (context->scan_cache->cache_last_fix_page);
+
+	      pgbuf_ordered_unfix_and_save_for_refix (thread_p, &context->scan_cache->page_watcher,
+						      saved_vpid, saved_latch_mode);
+	      assert (!VPID_ISNULL (&saved_vpid));
+	      assert (PGBUF_LATCH_READ == saved_latch_mode);
+	      restore_scan_cache_page_watcher = true;
+	    }
 	}
+
+      assert (PGBUF_IS_CLEAN_WATCHER (&context->home_page_watcher));
+      assert (PGBUF_IS_CLEAN_WATCHER (&context->fwd_page_watcher));
+      assert (PGBUF_IS_CLEAN_WATCHER (&context->scan_cache->page_watcher));
+
+      // wait for replication to catch-up
+      //
       constexpr VPID null_vpid = VPID_INITIALIZER;
       pgbuf_wait_for_replication (thread_p, &null_vpid);
+
+      // re-fix scan cache page watcher, if unfixed previously
+      //
+      if (restore_scan_cache_page_watcher)
+	{
+	  const int error_code = pgbuf_ordered_fix (thread_p, &saved_vpid, OLD_PAGE_MAYBE_DEALLOCATED,
+						    saved_latch_mode, &context->scan_cache->page_watcher);
+	  if (error_code != NO_ERROR)
+	    {
+	      // most likely, this [client read-only] transaction is expecting another thread
+	      // (most likely, the transactional log replication thread) to advance past a
+	      // certain point;
+	      // during replication, the page might be de-allocated while processing a log record
+	      // thus, failing to re-fix is expected
+	      //
+	      // TODO: in effect, the fetch mode is not correctly propagated to the page server;
+	      // instead - OLD_PAGE - is actually propagated to the
+	      // server (see pgbuf_request_data_page_from_page_server)
+	      // thus, if page has been de-allocated, an error will be retrieved from page server
+	      assert (false);
```

Comment:
Q: the debug version has this assert - does this means we think it should not happen, or is temporary to catch unhandled cases ?
In case of release version, do we continue ? I expected to break loop and return error.

Reply:
If the page cannot be re-fixed the single known cause is that the page has been de-allocated in the meantime. And since the only thing that executes changes on pages on PTS is the transactional log replication, this means that the page has been de-allocated and the replication has processed a `RVPGBUF_DEALLOC` for the page.

With the patch #3960, on PTS, once the `RVPGBUF_DEALLOC` log record is replicated, the page is invalidated from the page buffer (this might be an intermediary brute force solution - but it's what we have currently).
Thus, in such a scenario, if the page would be re-attempted for a fix at this point, the request for fix would transform into a request for fetch from Page Server, which will return an error code (**) as the page has been de-allocated.
In this case the assert is problably incorrect.
I will add a dump to err_log.
A subsequent interation - call to `heap_get_visible_version_internal` will probably come up empty handed and return.

(**)
There is even a worse case when the de-allocated page has already been re-allocated - maybe by another transaction on ATS. In which case, the fix will succeed but the layout of the page will be different and the search will fail - maybe even downright crash.
For this, we don't have an implementation. We might have a solution - described in http://jira.cubrid.org/browse/LETS-523 .

So, in all, I added the assert with the intention of catching one/both of the above described situations.